executor/src/witgen/data_structures/finalizable_vector.rs

```diff
@@ -0,0 +1,174 @@
+use std::ops::{Index, IndexMut};
+
+use ast::analyzed::PolyID;
+use bit_vec::BitVec;
+use number::FieldElement;
+
+use crate::witgen::rows::Row;
+
+trait FinalizableRow<T: FieldElement> {
+    fn finalize(&self, columns: &[PolyID]) -> Vec<T>;
+}
+
+impl<'a, T: FieldElement> FinalizableRow<T> for Row<'a, T> {
+    fn finalize(&self, columns: &[PolyID]) -> Vec<T> {
+        columns
+            .iter()
+            .map(|c| self[c].value.unwrap_or_default())
+            .collect()
+    }
+}
+
+#[derive(Clone)]
+enum Entry<'a, T: FieldElement> {
+    InProgress(Row<'a, T>),
+    Finalized(Vec<T>),
+}
+
+#[derive(Clone)]
+pub struct FinalizableData<'a, T: FieldElement> {
+    data: Vec<Entry<'a, T>>,
+    known_cells: Vec<BitVec>,
```

Comment:
This is only relevant for finalized rows, right? So it could move to Entry::Finalized